airbyte-integrations/connectors/destination-bigquery-denormalized/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java

```diff
@@ -0,0 +1,287 @@
+    List<JsonNode> valuesNode = node.findValues(attributeName);
+    Set<String> resultSet = new HashSet<>();
+    valuesNode.forEach(jsonNode -> {
+      if (jsonNode.isArray()) {
+        jsonNode.forEach(arrayNodeValue -> resultSet.add(arrayNodeValue.textValue()));
+      } else if (jsonNode.isObject()) {
+        resultSet.addAll(extractJsonValues(jsonNode, "value"));
+      } else {
+        resultSet.add(jsonNode.textValue());
+      }
+    });
+
+    return resultSet;
+  }
+
+  private List<JsonNode> retrieveRecordsAsJson(String tableName) throws Exception {
+    QueryJobConfiguration queryConfig =
+        QueryJobConfiguration
+            .newBuilder(
+                String.format("select TO_JSON_STRING(t) as jsonValue from %s.%s t;", dataset.getDatasetId().getDataset(), tableName.toLowerCase()))
+            .setUseLegacySql(false).build();
+
+    BigQueryUtils.executeQuery(bigquery, queryConfig);
+
+    return StreamSupport
+        .stream(BigQueryUtils.executeQuery(bigquery, queryConfig).getLeft().getQueryResults().iterateAll().spliterator(), false)
+        .map(v -> v.get("jsonValue").getStringValue())
+        .map(Jsons::deserialize)
+        .collect(Collectors.toList());
+  }
+
+  private JsonNode getSchema() {
+    try {
+      return new ObjectMapper().readTree(
+          "{\n"
+              + "  \"type\": [\n"
+              + "    \"object\"\n"
+              + "  ],\n"
+              + "  \"properties\": {\n"
+              + "    \"name\": {\n"
+              + "      \"type\": [\n"
+              + "        \"string\"\n"
+              + "      ]\n"
+              + "    },\n"
+              + "    \"permissions\": {\n"
+              + "      \"type\": [\n"
+              + "        \"array\"\n"
+              + "      ],\n"
+              + "      \"items\": {\n"
+              + "        \"type\": [\n"
+              + "          \"object\"\n"
+              + "        ],\n"
+              + "        \"properties\": {\n"
+              + "          \"domain\": {\n"
+              + "            \"type\": [\n"
+              + "              \"string\"\n"
+              + "            ]\n"
+              + "          },\n"
+              + "          \"grants\": {\n"
+              + "            \"type\": [\n"
+              + "              \"array\"\n"
+              + "            ],\n"
+              + "            \"items\": {\n"
+              + "              \"type\": [\n"
+              + "                \"string\"\n"
+              + "              ]\n"
+              + "            }\n"
+              + "          }\n"
+              + "        }\n"
+              + "      }\n"
+              + "    }\n"
+              + "  }\n"
+              + "}");
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static JsonNode getData() {
+    try {
+      return new ObjectMapper().readTree(
```

Comment:
we hide all Json operations behind an interface `Jsons.deserialize` to avoid spreading the exact json lib we're using in the code where possible